File: src/engine/adapters/sqs/definition.ts
import { GetQueueAttributesCommand, type SQSClient } from '@aws-sdk/client-sqs';
import type { DefinitionSection, ObjectDefinition } from '../../../shared/domain/definition';
import { encodePath } from '../../../shared/domain/tree';
import { mapError } from './errors';

// P23 D9: a queue genuinely *is* its attributes — visibility timeout, retention, delay, redrive
// policy, FIFO/dedup, KMS key, ARN, timestamps — none of which the app shows anywhere today. One
// command the adapter already imports (sqs/read.ts), one round trip, no new dependency. This is
// not a ReceiveMessage call and hides nothing — SPEC §5.1's "SQS reads are never automatic" rule
// is about message visibility, which GetQueueAttributes neither touches nor needs to auto-poll for.

// A few attribute values are themselves JSON policy documents — pretty-print those specifically
// rather than every attribute, so a plain string like an ARN or a timestamp stays a plain string.
const JSON_ATTRIBUTES = new Set(['RedrivePolicy', 'Policy', 'RedriveAllowPolicy']);

function formatValue(name: string, value: string): string {
  if (!JSON_ATTRIBUTES.has(name)) return value;
  try {
    return JSON.stringify(JSON.parse(value), null, 2);
  } catch {
    return value;
  }
}

export async function buildQueueDefinition(
  client: SQSClient,
  queueUrl: string,
  name: string,
): Promise<ObjectDefinition> {
  let attributes: Record<string, string>;
  try {
    const result = await client.send(
      new GetQueueAttributesCommand({ QueueUrl: queueUrl, AttributeNames: ['All'] }),
    );
    attributes = result.Attributes ?? {};
  } catch (err) {
    throw mapError(err);
  }

  const names = Object.keys(attributes).sort();
  const attributesSection: DefinitionSection = {
    title: 'Attributes',
    rows: names.map((n) => ({
      name: n,
      value: formatValue(n, attributes[n] as string),
      detail: null,
    })),
  };

  return {
    path: encodePath([{ kind: 'queue', name }]),
    kind: 'queue',
    qualifiedName: name,
    language: 'json',
    statements: [JSON.stringify(attributes, null, 2)],
    origin: 'server',
    notes: [],
    constraints: [],
    documentSchema: null,
    sections: [attributesSection],
    generatedAt: new Date().toISOString(),
  };
}

File: shell/internal/adapters/sqs/definition.go
package sqs

import (
	"bytes"
	"context"
	"encoding/json"
	"sort"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go-v2/service/sqs/types"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// jsonAttributes mirrors definition.ts's JSON_ATTRIBUTES — a few attribute values are themselves
// JSON policy documents; pretty-print those specifically rather than every attribute, so a plain
// string like an ARN or a timestamp stays a plain string.
var jsonAttributes = map[string]bool{"RedrivePolicy": true, "Policy": true, "RedriveAllowPolicy": true}

func formatAttributeValue(name, value string) string {
	if !jsonAttributes[name] {
		return value
	}
	var buf bytes.Buffer
	if err := json.Indent(&buf, []byte(value), "", "  "); err != nil {
		return value // swallow a parse failure back to the raw string, mirroring definition.ts
	}
	return buf.String()
}

// buildQueueDefinition is definition.ts's buildQueueDefinition — P23 D9: a queue genuinely *is*
// its attributes. One GetQueueAttributes(All) call, no automatic message read.
func buildQueueDefinition(ctx context.Context, client *sqs.Client, queueURL, name string) (model.ObjectDefinition, error) {
	result, err := client.GetQueueAttributes(ctx, &sqs.GetQueueAttributesInput{
		QueueUrl:       aws.String(queueURL),
		AttributeNames: []types.QueueAttributeName{types.QueueAttributeNameAll},
	})
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}
	attributes := result.Attributes

	names := make([]string, 0, len(attributes))
	for n := range attributes {
		names = append(names, n)
	}
	sort.Strings(names)

	rows := make([]model.DefinitionSectionRow, len(names))
	for i, n := range names {
		rows[i] = model.DefinitionSectionRow{Name: n, Value: formatAttributeValue(n, attributes[n]), Detail: nil}
	}

	// json.MarshalIndent of a map[string]string sorts keys — matching Object.keys().sort() above
	// for the rows, but not JS's JSON.stringify(attributes, null, 2) insertion order for this
	// statement text (P58d §4.2's recorded divergence; tests/ipc/sqs/sqs.fixture.ts freezes this
	// text against the TypeScript producer, and P58f's generator port will re-derive it).
	statementBytes, err := json.MarshalIndent(attributes, "", "  ")
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}

	return model.ObjectDefinition{
		Path:           model.EncodePath([]model.PathSegment{{Kind: "queue", Name: name}}),
		Kind:           "queue",
		QualifiedName:  name,
		Language:       "json",
		Statements:     []string{string(statementBytes)},
		Origin:         "server",
		Notes:          []string{},
		Constraints:    []model.ConstraintMeta{},
		DocumentSchema: nil,
		Sections:       []model.DefinitionSection{{Title: "Attributes", Rows: rows}},
		GeneratedAt:    time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}
